### PR TITLE
planner: don't DoOptimze when build show

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

A quick change to make `buildShow` in plan building more reasonable.

### What is changed and how it works?

Put it as a source of the `TableDual`.

### Check List 

Tests 

 - existing Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

### CODE REVIEW

File: planner/core/logical_plans.go [hunk 1]
```diff
@@ -313,10 +313,8 @@ type LogicalTableDual struct {
 	logicalSchemaProducer
 
 	RowCount int
-	// placeHolder indicates if this dual plan is a place holder in query optimization
-	// for data sources like `Show`, if true, the dual plan would be substituted by
-	// `Show` in the final plan.
-	placeHolder bool
+
+	sourcePlan Plan
```
Comment (francis0407) [inline planner/core/logical_plans.go]: Maybe add a comment here?

File: planner/core/logical_plan_test.go [hunk 2]
```diff
@@ -938,7 +938,7 @@ func (s *testPlanSuite) TestPlanBuilder(c *C) {
 		},
 		{
 			sql:  "show columns from t where `Key` = 'pri' like 't*'",
-			plan: "Show->Sel([eq(cast(key), 0)])",
+			plan: "Dual->Sel([eq(cast(key), 0)])->Projection",
```
Comment (eurekaka) [inline planner/core/logical_plan_test.go]: This plan generated looks more weird than the old one?

File: planner/core/common_plans.go [hunk 3]
```diff
@@ -368,6 +368,7 @@ type Deallocate struct {
 }
 
 // Show represents a show plan.
+// TODO: make as a logical data source or table dual.
```
Comment (lamxTyler) [inline planner/core/common_plans.go]: Why not just do it now?

Comment (winoros) [reply planner/core/common_plans.go]: i want to fix the unit-tests in my hands ASAP previously. :joy:

File: planner/core/logical_plan_test.go (hunk 2 above)
Comment (winoros) [reply planner/core/logical_plan_test.go]: Changed. But since the show doesn't do pruningm, the project is remained.

File: executor/compiler.go [hunk 4]
```diff
@@ -30,7 +30,7 @@ import (
 
 var (
 	stmtNodeCounterUse      = metrics.StmtNodeCounter.WithLabelValues("Use")
-	stmtNodeCounterShow     = metrics.StmtNodeCounter.WithLabelValues("Show")
+	stmtNodeCounterShow     = metrics.StmtNodeCounter.WithLabelValues("LogicalShow")
```
Comment (lamxTyler) [inline executor/compiler.go]: `Show` is better?

File: planner/core/planbuilder.go [hunk 5]
```diff
@@ -1394,38 +1394,23 @@ func (b *PlanBuilder) buildShow(ctx context.Context, show *ast.ShowStmt) (Plan,
...
 			schema.Append(newCol)
 		}
 		proj.SetSchema(schema)
 		proj.SetChildren(np)
-		physical, err := DoOptimize(ctx, b.optFlag|flagEliminateProjection, proj)
-		if err != nil {
-			return nil, err
-		}
-		return substitutePlaceHolderDual(physical, p), nil
+		return proj, nil
```
Comment (lzmhhh123) [inline planner/core/planbuilder.go]: Do we need to eliminate projection here?

File: planner/core/logical_plans.go [hunk 6]
```diff
@@ -765,3 +762,24 @@ func extractCorColumnsBySchema(p LogicalPlan, schema *expression.Schema) []*expr
...
+	Tp          ast.ShowStmtType // Databases/Tables/Columns/....
+	DBName      string
+	Table       *ast.TableName  // Used for showing columns.
+	Column      *ast.ColumnName // Used for `desc table column`.
+	IndexName   model.CIStr
+	Flag        int                  // Some flag parsed from sql, such as FULL.
+	User        *auth.UserIdentity   // Used for show grants.
+	Roles       []*auth.RoleIdentity // Used for show grants.
+	Full        bool
+	IfNotExists bool // Used for `show create database if not exists`
```
Comment (francis0407) [inline planner/core/logical_plans.go]: add `.` at the end of the comment.

File: planner/core/logical_plans.go [hunk 7]
```diff
@@ -765,3 +762,24 @@ func extractCorColumnsBySchema(p LogicalPlan, schema *expression.Schema) []*expr
...
+
+	GlobalScope bool // Used by show variables
```
Comment (francis0407) [inline planner/core/logical_plans.go]: ditto

File: planner/core/stats.go [hunk 8]
```diff
@@ -41,6 +41,19 @@ func (p *LogicalTableDual) DeriveStats(childStats []*property.StatsInfo) (*prope
 	return p.stats, nil
 }
 
+// DeriveStats implement LogicalPlan DeriveStats interface.
+func (p *LogicalShow) DeriveStats(childStats []*property.StatsInfo) (*property.StatsInfo, error) {
+	profile := &property.StatsInfo{
+		RowCount:    1,
```
Comment (eurekaka) [inline planner/core/stats.go]: Add a comment to explain the reason?

File: planner/core/planbuilder.go (hunk 5 above)
Comment (winoros) [reply planner/core/planbuilder.go]: Prvious one is not a valid project elimination.
For now, show cannot remove any of its output column.
So the project should be remained.

File: planner/core/stats.go [hunk 9]
```diff
@@ -41,6 +41,20 @@ func (p *LogicalTableDual) DeriveStats(childStats []*property.StatsInfo) (*prope
 	return p.stats, nil
 }
 
+// DeriveStats implement LogicalPlan DeriveStats interface.
```
Comment (francis0407) [inline planner/core/stats.go]: implement -> implements

File: planner/core/planbuilder.go (hunk 5 above)
Comment (eurekaka) [reply planner/core/planbuilder.go]: > Prvious one is not a valid project elimination.
For now, show cannot remove any of its output column.
So the project should be remained.

Sorry I didn't understand this, please explain a little bit more. Also, shall we set `flagEliminateProjection` of `b.optFlag` here?

Comment (winoros) [reply planner/core/planbuilder.go]: Oh, sorry i misread the comment.
We can add flag here.
But the project in the that test won't be removed

**Pull request merged**